### PR TITLE
Fix FormatException if an invalid JsonConverter attribute is specified

### PR DESCRIPTION
**Pull request by steveharter**
Replacement for PR https://github.com/dotnet/corefx/pull/39778 that is now targeted against master.

cc @martincostello 

cc @ericstj requesting approval to merge into 3.0 once in master.

### CODE REVIEW

File: src/System.Text.Json/tests/Serialization/CustomConverterTests.BadConverters.cs [hunk 1]
```diff
@@ -76,8 +77,102 @@ private class PocoWithNullConverter
...
+        public static void AttributeOnClassFail()
+        {
+            try
+            {
+                JsonSerializer.Serialize(new InvalidTypeConverterClassWithAttribute());
+                Assert.True(false, "Expected InvalidOperationException was not thrown.");
+            }
+            catch (InvalidOperationException ex)
+            {
+                // Message should be in the form "The converter specified on 'System.Text.Json.Serialization.Tests.CustomConverterTests+InvalidTypeConverterClassWithAttribute' is not compatible with the type 'System.Text.Json.Serialization.Tests.ExceptionTests+InvalidTypeConverterClassWithAttribute'."
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/CustomConverterTests.BadConverters.cs]: Is there a comment mismatch?
Where is "ExceptionTests+InvalidTypeConverterClassWithAttribute" coming from?

File: src/System.Text.Json/tests/Serialization/CustomConverterTests.BadConverters.cs [hunk 2]
```diff
@@ -76,8 +77,102 @@ private class PocoWithNullConverter
         [Fact]
         public static void AttributeCreateConverterFail()
         {
-            Assert.Throws<InvalidOperationException>(() => JsonSerializer.Serialize(new PocoWithNullConverter()));
-            Assert.Throws<InvalidOperationException>(() => JsonSerializer.Deserialize<PocoWithNullConverter>("{}"));
+            try
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/CustomConverterTests.BadConverters.cs]: I don't think changing to try/catch is necessary. We can keep using Assert.Throws, which returns the exception.
```C#
InvalidOperationException ex = Assert.Throws<InvalidOperationException>(() => JsonSerializer.Serialize(new PocoWithNullConverter()));
Assert.Contains("'System.Text.Json.Serialization.Tests.CustomConverterTests+PocoWithNullConverter.MyInt'", ex.Message);
```

File: src/System.Text.Json/tests/Serialization/CustomConverterTests.BadConverters.cs [hunk 3]
```diff
@@ -76,8 +77,102 @@ private class PocoWithNullConverter
...
+                JsonSerializer.Serialize(new PocoWithNullConverter());
+                Assert.True(false, "Expected InvalidOperationException was not thrown.");
+            }
+            catch (InvalidOperationException ex)
+            {
+                // Message should be in the form: "The converter specified on 'System.Text.Json.Serialization.Tests.CustomConverterTests+PocoWithNullConverter.MyInt' does not derive from JsonConverter or have a public parameterless constructor."
+                Assert.Contains("'System.Text.Json.Serialization.Tests.CustomConverterTests+PocoWithNullConverter.MyInt'", ex.Message);
+            }
+
+            try
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/CustomConverterTests.BadConverters.cs]: Same here/elsewhere.

File: src/System.Text.Json/tests/Serialization/CustomConverterTests.BadConverters.cs (hunk 1 above)
Comment (steveharter) [reply src/System.Text.Json/tests/Serialization/CustomConverterTests.BadConverters.cs]: Thanks yes updated that comment.

File: src/System.Text.Json/tests/Serialization/CustomConverterTests.BadConverters.cs (hunk 2 above)
Comment (steveharter) [reply src/System.Text.Json/tests/Serialization/CustomConverterTests.BadConverters.cs]: I never realized that. I'll clean this file up.

**Comment by steveharter** [comment]
Test failure (MacOS x64_Debug) not related. Infrastructure Execution Logs on:
System.Diagnostics.Contracts.Tests

File: src/System.Text.Json/tests/Serialization/CustomConverterTests.BadConverters.cs [hunk 4]
```diff
@@ -57,27 +58,78 @@ public static void ContraVariantConverterFail()
...
+            Value2,
+        }
+
+        [Fact]
+        public static void AttributeOnPropertyFail()
+        {
+            InvalidOperationException ex;
+
+            ex = Assert.Throws<InvalidOperationException>(() => JsonSerializer.Serialize(new InvalidTypeConverterClass()));
+            // Message should be in the form "The converter specified on 'System.Text.Json.Serialization.Tests.CustomConverterTests+InvalidTypeConverterClass.MyEnumValues' is not compatible with the type 'System.Collections.Generic.ICollection`1[System.Text.Json.Serialization.Tests.ExceptionTests+InvalidTypeConverterEnum]'."
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/CustomConverterTests.BadConverters.cs]: This comment is still not correct. We don't have `**ExceptionTests**+InvalidTypeConverterEnum`.

Shouldn't this be:
`System.Text.Json.Serialization.Tests.CustomConverterTests+InvalidTypeConverterEnum`

**Comment by karelz** [comment]
This PR went into master - fixing milestone to 5.0.
It was later ported into release/3.0 branch in PR #39811.

**Pull request merged**